File: org.opencompare/io-wikipedia/src/main/scala/org/opencompare/io/wikipedia/pcm/Matrix.scala
package org.opencompare.io.wikipedia.pcm

import org.opencompare.api.java.PCM

import scala.xml.Elem
import scala.xml.Text

class Matrix {

  var name : String = ""
  var cells : Map[(Int, Int), Cell] = Map()
  
  def setCell(cell : Cell, row : Int, column : Int) {
    cells += ((row, column) -> cell)
  }
  
  def getCell(row : Int, column : Int) : Option[Cell] = {
    cells get (row, column)
  }
  
  def getNumberOfRows() : Int = {
    if (!cells.isEmpty) {
    	cells.keys.maxBy(_._1)._1 + 1
    } else {
      0
    }
  }
  
  def getNumberOfColumns() : Int = {
    if (!cells.isEmpty) {
    	cells.keys.maxBy(_._2)._2 + 1
    } else {
      0
    }
  }
  
  override def toString() : String = {
    val result = new StringBuilder
    for (row <- 0 until getNumberOfRows; column <- 0 until getNumberOfColumns) {
      result ++= row + "," + column + ":"
      
      val cell = cells get (row, column)
      if (cell.isDefined) {
        result ++= cell.get.content 
      } else {
        result ++= "/!\\ This cell is not defined /!\\"
      }
      result += '\n'
    }
    result.toString
  }
  
  def toHTML() : Elem = {
    val htmlCode = 
    <div>
    <h1>{name}</h1>
    <table border="1">
    { 
    	for {row <- 0 until getNumberOfRows} 
	    yield <tr> 
      	  {
		    for {column <- 0 until getNumberOfColumns} 
		  	yield <td>
		  	{
		  	  val cell = cells.get((row, column))
		  	  if (cell.isDefined) {
		  	    // Convert new lines in <br/>
		  	    val lines = cell.get.content.split("\n")
		  	    var firstLine = true
		  	    for (line <- lines) yield {
		  	      if (firstLine) {
		  	        firstLine = false
		  	        Text(line)
		  	      } else {
		  	        <br/> ++ Text(line)
		  	      }
		  	    }
		  	  } else {
		  	    "/!\\ Not defined /!\\"
		  	  }
		  	}
		  	</td>
		  }
    	</tr>
	} 
    </table>
    </div>
    
    htmlCode
  }

  def comparePosition(c1 : Cell, c2 : Cell) : Boolean = {
    (c1.row < c2.row) || 
    (c1.row == c2.row && c1.column < c2.column)
  }
  
  def toCSV() : String = {
    val result = new StringBuilder
    for (row <- 0 until getNumberOfRows) {
      for (column <- 0 until getNumberOfColumns) {
    	  val cell = cells get (row, column)
    	  if (cell.isDefined) {
    	    val content = cell.get.content
    	    val formattedContent = '"' + content.replaceAll("\"", "\"\"") + '"'
    		result ++= formattedContent 
    	  } else {
    		result ++= "\"/!\\ This cell is not defined /!\\\""
    	  }
    	  if (column < getNumberOfColumns - 1) {
    	    result += ','
    	  }
      }
      
      if (row < getNumberOfRows - 1) {
    	  result += '\n'  
      }
      
    }
    result.toString
  }
  
  def toPCM(pcm : PCM) {
    val nbFeatureLines = 1
    val nbProductColumns = 1


    // TODO : set first line (+ followings if rowspan) as features
    // TODO : set first column (+ followings if colspan) as products

  }
}

File: org.opencompare/api-java/src/main/java/org/opencompare/api/java/io/Matrix.java
package org.opencompare.api.java.io;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by smangin on 02/07/15.
 */
public class Matrix {

    public String name = "";
    public Map<Map<Integer, Integer>, Cell> cells = new HashMap<>();

    public Matrix(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setCell(Cell cell, int row, int column) {
        cells.put(new HashMap<Integer, Integer>(row, column), cell);
    }

    public Cell getCell(int row, int column) {
        return cells.get(new HashMap<Integer, Integer>(row, column));
    }

    public int getNumberOfRows() {
        if (!cells.isEmpty()) {
            return cells.keySet().size();
        }
        return 0;
    }

    public int getNumberOfColumns() {
        if (!cells.isEmpty()) {
            return cells.keySet().size();
        }
        return 0;
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < getNumberOfRows(); i++){
            for (int j = 0; j < getNumberOfColumns(); j++) {
                result.append(i + "," + j + ":");

                Cell cell = cells.get(new HashMap<Integer, Integer>(i, j));
                if (cell != null) {
                    result.append(cell.getContent());
                } else {
                    result.append("/!\\ This cell is not publicined /!\\");
                }
                result.append('\n');
            }
        }
        return result.toString();
    }


    public boolean comparePosition(Cell c1, Cell c2) {
        return (c1.getRow() < c2.getRow()) || (c1.getRow() == c2.getRow() && c1.getColumn() < c2.getColumn());
    }

}
